### PR TITLE
devices - move test utils to their own public module(s)

### PR DESCRIPTION
**Pull request by AlexandruCihodaru**
Signed-off-by: Alexandru Cihodaru <cihodar@amazon.com>

## Reason for This PR

Refactored the code from devices crate such that the utility functions and structures are now in public mods. The changes are necessary so Firecracker could be integrated easier with other projects.

## Description of Changes

[devices] expose helper functions as public.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

**Comment by raduiliescu** [comment]
Alex, can you please put a better title and change description. You can describe why you needed the refactoring.

File: src/devices/src/virtio/block/device.rs [hunk 1]
```diff
@@ -412,43 +424,16 @@ impl VirtioDevice for Block {
...
-    use std::os::unix::io::AsRawFd;
-    use std::thread;
-    use std::time::Duration;
-    use std::u32;
-
-    use super::*;
-    use crate::virtio::queue::tests::*;
-    use polly::event_manager::{EventManager, Subscriber};
-    use utils::epoll::{EpollEvent, EventSet};
+pub mod block_utils {
```
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:427]: All of these test utils that need to be used outside of unit-tests (I am guessing in your case they'd be used in fuzzing), should go to a module uniformly called `test_utils` across crates.

In virtio-case these utils will be smth like:
```rust
use device::virtio::queue::test_utils::*;
use device::virtio::block::test_utils::*;
use device::virtio::net::test_utils::*;
// etc.
```

Naming them uniformly makes it easy to find them and to also exclude them from coverage same as we exclude `tests` modules.

**Review by acatangiu — COMMENTED** [review]
The PR and commit descriptions should be descriptive. Please describe what problem you are tackling and how you are solving it.

File: src/devices/src/virtio/block/device.rs (hunk 1 above)
Comment (acatangiu) [reply src/devices/src/virtio/block/device.rs:427]: Ideally they could even live in their own `test_utils.rs` subfiles.

File: src/devices/src/virtio/block/block_test_utils.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+pub mod test_utils {
```
Comment (acatangiu) [inline src/devices/src/virtio/block/block_test_utils.rs:1]: With current code you have the following hierarchy:
`devices::virtio::block::block_test_utils::test_utils`

You can simplify this by renaming the file from `block_test_utils.rs` to `test_utils.rs` since using a sibling file next to `mod.rs` implies all code in that file being in a module with the same name as the filename.

The hierarchy would then become: `devices::virtio::block::test_utils` which is simpler and nicer.

Same for all others.

File: src/devices/src/virtio/net/device.rs [hunk 3]
```diff
@@ -704,6 +708,84 @@ impl Net {
             }
         }
     }
+
+    pub fn default_net() -> Net {
```
Comment (acatangiu) [inline src/devices/src/virtio/net/device.rs:712]: This should go in `net::test_utils`

File: src/devices/src/virtio/block/device.rs [hunk 4]
```diff
@@ -329,6 +329,18 @@ impl Block {
...
+    pub fn set_queue(&mut self, idx: usize, q: Queue) {
+        self.queues[idx] = q;
+    }
+
+    pub fn set_rate_limiter(&mut self, rl: RateLimiter) {
+        self.rate_limiter = rl;
+    }
+
+    pub fn rate_limiter(&mut self) -> &RateLimiter {
+        &self.rate_limiter
```
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:342]: Since `self.queues` is `pub(crate)` and these setters and getters are only used in tests, I propose you move their definition to `block::test_utils`

File: src/devices/src/virtio/net/tap.rs [hunk 5]
```diff
@@ -163,6 +166,40 @@ impl Tap {
...
+        IfReqBuilder::new()
+            .if_name(&self.if_name)
+            .flags(
+                (net_gen::net_device_flags_IFF_UP
+                    | net_gen::net_device_flags_IFF_RUNNING
+                    | net_gen::net_device_flags_IFF_NOARP) as i16,
+            )
+            .execute(&sock, c_ulong::from(net_gen::sockios::SIOCSIFFLAGS))
+            .unwrap();
+    }
```
Comment (acatangiu) [inline src/devices/src/virtio/net/tap.rs:202]: Not sure if straightforward but these functions should definitely go in `net::test_utils` somehow.
They definitely don't make sense as `firecracker` binary code.

File: src/devices/src/virtio/net/tap.rs [hunk 6]
```diff
@@ -203,134 +238,7 @@ pub mod tests {
...
-                    &mut (mem::size_of::<libc::sockaddr_storage>() as libc::socklen_t),
-                )
-            };
-            if ret == -1 {
-                return false;
-            }
-            true
-        }
-    }
+    impl Tap {}
```
Comment (acatangiu) [inline src/devices/src/virtio/net/tap.rs:241]:
```suggestion
```

File: src/devices/src/virtio/queue.rs [hunk 7]
```diff
@@ -384,268 +384,16 @@ impl Queue {
     }
 }
 
+pub mod queue_utils {}
```
Comment (acatangiu) [inline src/devices/src/virtio/queue.rs:387]:
```suggestion
```

File: src/devices/src/virtio/vsock/device.rs [hunk 8]
```diff
@@ -217,6 +219,25 @@ where
...
+    }
+
+    pub fn get_element_from_queue(&self, idx: usize) -> result::Result<u64, DeviceError> {
+        if idx > self.queue_events.len() {
+            return Err(DeviceError::QueueError(QueueError::DescIndexOutOfBounds(
+                idx as u16,
+            )));
+        }
+        Ok(self.queue_events[idx].as_raw_fd() as u64)
+    }
```
Comment (acatangiu) [inline src/devices/src/virtio/vsock/device.rs:240]: Would be nice if we could have this in `vsock::test_utils` since they're not used in release code.

File: tests/integration_tests/build/test_coverage.py [hunk 9]
```diff
@@ -23,7 +23,7 @@
 # this contains the frequency while on AMD it does not.
 # Checkout the cpuid crate. In the future other
 # differences may appear.
-COVERAGE_DICT = {"Intel": 85.08, "AMD": 84.28}
+COVERAGE_DICT = {"Intel": 85.34, "AMD": 84.58}
```
Comment (acatangiu) [inline tests/integration_tests/build/test_coverage.py:26]: Once all these utils go to their respective `test_utils.rs` files, you can update the `exclude_pattern` in `tests/integration_tests/build/test_coverage.py` to exclude this code from coverage.

Our production code coverage should not include test utils.

**Review by acatangiu — CHANGES_REQUESTED** [review]
Please add more detail to your commit message. You can actually explain that there are utility funcs defined in `mod tests` and these are not accessible outside the crate, and to use them in other crates they need to be made public. That you chose `test_utils.rs` for them, how you structure them, etc.

Also follow the [50/72 rule](https://medium.com/@preslavrachev/what-s-with-the-50-72-rule-8a906f61f09c) when formatting your commit message.

File: tests/integration_tests/build/test_coverage.py [hunk 10]
```diff
@@ -97,6 +97,7 @@ def test_coverage(test_session_root_path, test_session_tmp_path):
         'tests/,'
         'usr/lib/gcc,'
         'lib/x86_64-linux-gnu/,'
+        'test_utils.rs'
```
Comment (acatangiu) [inline tests/integration_tests/build/test_coverage.py:100]:
```suggestion
        'test_utils.rs,'
```

File: src/devices/src/virtio/net/device.rs [hunk 11]
```diff
@@ -122,8 +124,7 @@ pub struct Net {
 
     pub(crate) mmds_ns: Option<MmdsNetworkStack>,
 
-    #[cfg(test)]
-    mocks: tests::Mocks,
+    pub(crate) mocks: Mocks,
```
[the remark this answers is not in the document]
Comment (serban300) [reply src/devices/src/virtio/net/device.rs:127]: Do we absolutely need to access `mocks` from outside this create ? If we do, can we provide it as a function parameter ?

File: src/devices/src/virtio/balloon/device.rs [hunk 12]
```diff
@@ -465,6 +465,7 @@ pub(crate) mod tests {
     use super::super::CONFIG_SPACE_SIZE;
     use super::*;
     use crate::virtio::queue::tests::*;
```
[the remark this answers is not in the document]
Comment (AlexandruCihodaru) [reply src/devices/src/virtio/balloon/device.rs]: I have removed bla::* and only imported used things.

File: src/devices/src/virtio/net/tap.rs (hunk 5 above)
Comment (AlexandruCihodaru) [reply src/devices/src/virtio/net/tap.rs]: I have moved them in net/test_utils.rs. Take a look please.

File: src/devices/src/virtio/net/device.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (AlexandruCihodaru) [reply src/devices/src/virtio/net/device.rs:128]: I chose to gate mocks with `#[cfg(test)]` and also moved TestHalper(the structure that used mocks field from net) into a test module in file net/test_utils.rs. Take a look please @serban300 @acatangiu

File: src/snapshot/Cargo.toml [hunk 13]
```diff
@@ -10,6 +10,6 @@ libc = "0.2"
 versionize = "0.1.2"
 versionize_derive = ">=0.1.1"
 
-#[[bench]]
-#name = "main"
-#harness = false
+[[bench]]
+name = "main"
+harness = false
```
Comment (acatangiu) [inline src/snapshot/Cargo.toml:15]: nit: Is this intended? If so, I think it's unrelated to the rest of the changes in this commit. Maybe move it to a separate commit.

**Review by acatangiu — APPROVED** [review]

**Review by acatangiu — DISMISSED** [review]

File: src/devices/src/virtio/block/test_utils.rs [hunk 14]
```diff
@@ -0,0 +1,51 @@
+use std::os::unix::io::AsRawFd;
+
+use polly::event_manager::{EventManager, Subscriber};
+use rate_limiter::RateLimiter;
+use utils::epoll::{EpollEvent, EventSet};
+use utils::tempfile::TempFile;
+
```
Comment (dianpopa) [inline src/devices/src/virtio/block/test_utils.rs:7]: nit: superfluous newline

File: src/devices/src/virtio/block/device.rs [hunk 15]
```diff
@@ -427,6 +427,11 @@ pub(crate) mod tests {
     use utils::tempfile::TempFile;
     use vm_memory::GuestAddress;
 
+    use crate::virtio::block::test_utils::{
```
Comment (dianpopa) [inline src/devices/src/virtio/block/device.rs:430]: nit: You could just use
`use crate:;virtio:;block::test_utils::*`. since we expect to define lots of stuff in these type of modules and use all of them

File: src/devices/src/virtio/balloon/device.rs [hunk 16]
```diff
@@ -464,7 +464,8 @@ pub(crate) mod tests {
 
     use super::super::CONFIG_SPACE_SIZE;
     use super::*;
-    use crate::virtio::queue::tests::*;
+    use crate::virtio::test_utils::VirtQueue;
```
Comment (dianpopa) [inline src/devices/src/virtio/balloon/device.rs:467]: Ballon device has some testing utilites that we may want to move in a separate module, right?

Comment (dianpopa) [reply src/devices/src/virtio/balloon/device.rs:467]: Pay attention that `default_mem` is already present in `virtio/test_utils`.

File: src/devices/src/virtio/vsock/device.rs [hunk 17]
```diff
@@ -24,6 +24,7 @@ use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 
 use logger::{debug, error, warn, Metric, METRICS};
+use std::os::unix::io::AsRawFd;
```
Comment (dianpopa) [inline src/devices/src/virtio/vsock/device.rs:27]: nit: move this after `use std::sync::Arc;`

File: src/devices/src/virtio/vsock/device.rs (hunk 8 above)
Comment (dianpopa) [reply src/devices/src/virtio/vsock/device.rs:240]: These functions were added in this PR?

File: src/devices/src/virtio/vsock/test_utils.rs [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+use crate::virtio::test_utils::VirtQueue as GuestQ;
+use crate::virtio::vsock::device::{RXQ_INDEX, TXQ_INDEX};
+use crate::virtio::vsock::packet::{VsockPacket, VSOCK_PKT_HDR_SIZE};
+use crate::virtio::{
+    VirtioDevice, Vsock, VsockBackend, VsockChannel, VsockEpollListener, VsockError,
+    VIRTQ_DESC_F_NEXT, VIRTQ_DESC_F_WRITE,
+};
+
+use vm_memory::{GuestAddress, GuestMemoryMmap};
+
```
Comment (dianpopa) [inline src/devices/src/virtio/vsock/test_utils.rs:12]: nit: no need for newline

File: src/devices/src/virtio/vsock/test_utils.rs [hunk 19]
```diff
@@ -0,0 +1,193 @@
+use std::os::unix::io::{AsRawFd, RawFd};
```
Comment (dianpopa) [inline src/devices/src/virtio/vsock/test_utils.rs:1]: copyright header

**Review by dianpopa — COMMENTED** [review]
1. The PR is very big and it is hard to review when all changes are placed in one commit. I would break up the PR in differents commits like for example:
* block: move testing utilities in separate module
 * net: move testing utilities in separate module
* balloon: move testing utilities in separate module
* vsock: move
* virtio: move testing utilities in separate module
* vsock: added helper function for ...
* CI: exclude testing utilities from coverage report
`
You can do these by using `git reset HEAD^` and then `git add -p`

2. Any functionality added on top of the existent one should reside in its own commit so that we can distinguish between it and functionality that was just relocated.

3. check_metric_after_block is redefined in all devices. Couldn't we move it in `virtio/test_utils`?

4.  New test_utils.rs files need to have the copyright message

File: src/devices/src/virtio/block/device.rs (hunk 15 above)
Comment (AlexandruCihodaru) [reply src/devices/src/virtio/block/device.rs:430]: https://github.com/firecracker-microvm/firecracker/pull/2151#discussion_r499448695 I was considering this when imported utilitary functions. Personally I consider that a good practice, let me know what you think.

Comment (dianpopa) [reply src/devices/src/virtio/block/device.rs:430]: It is up to you.

**Review by dianpopa — APPROVED** [review]
Nice job! 💯

**Review by acatangiu — APPROVED** [review]
